File: Python/8-Graphs(BFS-DFS)/Algorithms/6-TopologicalSorting(Kahn’sAlgorithm)/TopologicalSorting(Kahn’sAlgorithm)Pro.py
from collections import defaultdict, deque
from typing import List, Dict, Set, Tuple, Optional

# 1. Basic Kahn's Algorithm - Dependency Resolution (Build Systems)
def topological_sort_kahn_basic(graph: Dict[str, List[str]]) -> List[str]:
    """
    Implements Kahn's algorithm for topological sorting.

    Args:
        graph: A dictionary representing the directed acyclic graph (DAG).
               Keys are nodes, and values are lists of their dependent nodes.

    Returns:
        A list of nodes in topological order. Returns an empty list if a cycle exists.

    Example Application:
        In a software build system (like Make or CMake), tasks (nodes) depend on other tasks
        (e.g., compiling a source file depends on having the header files). Topological sort
        determines the order in which to execute the tasks.

    """
    in_degree = {node: 0 for node in graph}  # Calculate in-degrees
    for node in graph:
        for neighbor in graph[node]:
            in_degree[neighbor] += 1

    queue = deque([node for node in in_degree if in_degree[node] == 0])  # Start with nodes with no dependencies
    result = []

    while queue:
        node = queue.popleft()
        result.append(node)
        for neighbor in graph[node]:
            in_degree[neighbor] -= 1
            if in_degree[neighbor] == 0:
                queue.append(neighbor)

    if len(result) != len(graph):
        return []  # Cycle detected
    else:
        return result

# 2. Kahn's Algorithm with Cycle Detection - Task Scheduling with Dependencies
def topological_sort_kahn_cycle_detect(graph: Dict[str, List[str]]) -> Tuple[List[str], bool]:
    """
    Kahn's algorithm with explicit cycle detection.

    Args:
        graph: Adjacency list representation of the graph.

    Returns:
        A tuple:
            - The topological order (list of nodes) or an empty list if a cycle exists.
            - A boolean indicating whether a cycle was detected (True if cycle, False otherwise).

     Example Application:
        Task scheduling in project management.  Tasks have dependencies, and you need to schedule
        them in an order that satisfies those dependencies.  This version detects if the dependencies
        are contradictory (a cycle), which would mean the project is impossible to schedule.
    """
    in_degree = {node: 0 for node in graph}
    for node in graph:
        for neighbor in graph[node]:
            in_degree[neighbor] += 1

    queue = deque([node for node in in_degree if in_degree[node] == 0])
    result = []
    nodes_visited = 0 # Keep track of how many nodes were visited

    while queue:
        node = queue.popleft()
        result.append(node)
        nodes_visited += 1
        for neighbor in graph[node]:
            in_degree[neighbor] -= 1
            if in_degree[neighbor] == 0:
                queue.append(neighbor)

    if nodes_visited != len(graph):
        return [], True  # Cycle detected
    else:
        return result, False

# 3. Kahn's Algorithm with Priority - Course Scheduling with Prerequisites
def topological_sort_kahn_priority(graph: Dict[str, List[str]], priorities: Dict[str, int]) -> List[str]:
    """
    Kahn's algorithm, considering node priorities.

    Args:
        graph: Adjacency list representation of the graph.
        priorities: A dictionary mapping nodes to their priorities (lower value means higher priority).

    Returns:
        A list of nodes in topological order, considering priorities.  If there's a tie,
        nodes with higher priority (lower priority value) are added to the result first.

    Example Application:
        Course scheduling in a university.  Courses have prerequisites (dependencies),
        and some courses might have higher priority (e.g., required courses for a major).
        This version ensures that high-priority courses are scheduled as early as possible.
    """
    in_degree = {node: 0 for node in graph}
    for node in graph:
        for neighbor in graph[node]:
            in_degree[neighbor] += 1

    # Use a list and sort it to simulate a priority queue.  For large graphs, a heap would be more efficient.
    queue_list = [(node, in_degree[node]) for node in in_degree if in_degree[node] == 0]
    queue_list.sort(key=lambda x: priorities[x[0]])  # Sort by priority

    queue = deque(queue_list)
    result = []

    while queue:
        node, _ = queue.popleft()
        result.append(node)
        for neighbor in graph[node]:
            in_degree[neighbor] -= 1
            if in_degree[neighbor] == 0:
                queue_list.append((neighbor, in_degree[neighbor]))
                queue_list.sort(key=lambda x: priorities[x[0]]) # keep sorted
                queue = deque(queue_list) #update queue

    if len(result) != len(graph):
        return []  # Cycle detected
    else:
        return result
# 4. Kahn's Algorithm with Feedback - Dependency Resolution with Updates
def topological_sort_kahn_feedback(graph: Dict[str, List[str]], initial_order: Optional[List[str]] = None) -> List[str]:
    """
    Kahn's algorithm that can incorporate a partial initial ordering.

    Args:
        graph: Adjacency list representation of the graph.
        initial_order: An optional list representing a partial topological order.
                       Nodes in this list are prioritized in the sorting process.

    Returns:
        A topological ordering of the nodes.

    Example Application:
        Software dependency management where a user provides a partial order of installation.
        For example, a user might want a specific library to be installed first.  This version
        respects that preference while still satisfying dependencies.
    """
    in_degree = {node: 0 for node in graph}
    for node in graph:
        for neighbor in graph[node]:
            in_degree[neighbor] += 1

    queue = deque()
    if initial_order:
        for node in initial_order:
            if in_degree[node] == 0:
                queue.append(node)
                in_degree[node] = -1  # Mark as added to queue
        for node in in_degree: #add the rest
            if in_degree[node] == 0:
                queue.append(node)
    else:
        queue = deque([node for node in in_degree if in_degree[node] == 0])

    result = []

    while queue:
        node = queue.popleft()
        result.append(node)
        for neighbor in graph[node]:
            in_degree[neighbor] -= 1
            if in_degree[neighbor] == 0:
                queue.append(neighbor)

    if len(result) != len(graph):
        return []  # Cycle detected
    else:
        return result

# 5. Kahn's Algorithm for DAG Validation - Workflow Validation
def is_dag_kahn(graph: Dict[str, List[str]]) -> bool:
    """
    Uses Kahn's algorithm to determine if a graph is a DAG.  It's not strictly a topological sort,
    but it leverages the same logic.

    Args:
        graph: Adjacency list representation of the graph.

    Returns:
        True if the graph is a DAG, False otherwise.

    Example Application:
        Validating workflows in a business process management system.  A workflow can be represented
        as a graph where activities are nodes and dependencies between activities are edges.  This
        function checks if the workflow definition is valid (i.e., contains no circular dependencies).
    """
    in_degree = {node: 0 for node in graph}
    for node in graph:
        for neighbor in graph[node]:
            in_degree[neighbor] += 1

    queue = deque([node for node in in_degree if in_degree[node] == 0])
    count = 0

    while queue:
        node = queue.popleft()
        count += 1
        for neighbor in graph[node]:
            in_degree[neighbor] -= 1
            if in_degree[neighbor] == 0:
                queue.append(neighbor)

    return count == len(graph) #if count is equal to number of nodes, then it is a DAG

# 6. Kahn's Algorithm with Edge Weights (for shortest path in DAG) - Project Scheduling with Time Constraints
def topological_sort_kahn_weighted(graph: Dict[str, List[Tuple[str, int]]]) -> List[str]:
    """
    Kahn's algorithm for a graph with edge weights.  While not a pure topological sort,
    it's adapted to handle weighted edges, which is useful for finding the longest path
    in a DAG (which is related to critical path analysis).

    Args:
        graph:  Adjacency list where values are lists of (neighbor, weight) tuples.
                For example: {'A': [('B', 5), ('C', 3)], 'B': [('D', 2)]}
        start_node: The starting node for the longest path calculation.

    Returns:
        A topological ordering of the nodes.

    Example Application:
        Project scheduling with time constraints.  Nodes are tasks, edges are dependencies,
        and edge weights are the time it takes to complete a task.  This can be adapted to find
        the critical path (the longest sequence of tasks that determines the project duration).
    """
    in_degree = {node: 0 for node in graph}
    for node in graph:
        for neighbor, _ in graph[node]:  # Iterate through neighbors and weights, we only care about neighbors for in_degree
            in_degree[neighbor] += 1

    queue = deque([node for node in in_degree if in_degree[node] == 0])
    result = []

    while queue:
        node = queue.popleft()
        result.append(node)
        for neighbor, _ in graph[node]: # Iterate through neighbors and weights.
            in_degree[neighbor] -= 1
            if in_degree[neighbor] == 0:
                queue.append(neighbor)

    if len(result) != len(graph):
        return []  # Cycle detected
    else:
        return result

def main():
    """
    Main function to demonstrate the different topological sort implementations.
    """
    # Example graph (Directed Acyclic Graph)
    graph = {
        'A': ['B', 'C'],
        'B': ['D', 'E'],
        'C': ['F'],
        'D': ['E', 'F'],
        'E': ['F'],
        'F': []
    }

    print("1. Basic Kahn's Algorithm - Dependency Resolution:")
    result1 = topological_sort_kahn_basic(graph)
    print("   Topological Order:", result1)  # Output: ['A', 'C', 'B', 'D', 'E', 'F'] or other valid order

    print("\n2. Kahn's Algorithm with Cycle Detection - Task Scheduling:")
    result2, has_cycle = topological_sort_kahn_cycle_detect(graph)
    print("   Topological Order:", result2)
    print("   Cycle Detected:", has_cycle)  # Output: False

    # Example graph with a cycle
    graph_with_cycle = {
        'A': ['B'],
        'B': ['C'],
        'C': ['A']
    }
    result2_cycle, has_cycle_cycle = topological_sort_kahn_cycle_detect(graph_with_cycle)
    print("\n   With Cycle:")
    print("   Topological Order:", result2_cycle)
    print("   Cycle Detected:", has_cycle_cycle)  # Output: True

    print("\n3. Kahn's Algorithm with Priority - Course Scheduling:")
    priorities = {'A': 2, 'B': 1, 'C': 3, 'D': 1, 'E': 2, 'F': 3}  # Lower value = higher priority
    result3 = topological_sort_kahn_priority(graph, priorities)
    print("   Topological Order with Priorities:", result3)  # Output:  ['B', 'D', 'A', 'E', 'C', 'F']

    print("\n4. Kahn's Algorithm with Feedback - Dependency Resolution with Updates:")
    initial_order = ['B', 'C']  # User wants B and C to be processed first, if possible
    result4 = topological_sort_kahn_feedback(graph, initial_order)
    print("   Topological Order with Feedback:", result4) # Output: ['B', 'C', 'A', 'D', 'E', 'F']

    print("\n5. Kahn's Algorithm for DAG Validation - Workflow Validation:")
    is_dag = is_dag_kahn(graph)
    print("   Is DAG:", is_dag)  # Output: True
    is_dag_cycle = is_dag_kahn(graph_with_cycle)
    print("   Is DAG (with cycle):", is_dag_cycle) # Output: False

    print("\n6. Kahn's Algorithm with Edge Weights - Project Scheduling with Time Constraints:")
    weighted_graph = {
        'A': [('B', 5), ('C', 3)],
        'B': [('D', 2), ('E', 4)],
        'C': [('F', 1)],
        'D': [('E', 1), ('F', 6)],
        'E': [('F', 2)],
        'F': []
    }
    result6 = topological_sort_kahn_weighted(weighted_graph)
    print("  Topological Order with Weights:", result6)

if __name__ == "__main__":
    main()

File: C++/8-Graphs(BFS-DFS)/Algorithms/6-TopologicalSorting(Kahn’sAlgorithm)/TopologicalSorting(Kahn’sAlgorithm)Pro.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
#include <functional>
#include <unordered_map>

using namespace std;

// 1. Basic Kahn's Algorithm (Task Scheduling)
//
//   -   Problem:  You have a list of tasks to perform, and some tasks depend on others being completed first.  Find an order in which to perform the tasks.
//   -   Real-world:  This is the classic application of topological sort.  Think of a build system (like Make or CMake) where compiling a library depends on compiling its dependencies.
//   -   Code:
vector<int> topologicalSort_Kahn_Basic(int numTasks, const vector<pair<int, int>>& prerequisites) {
    vector<vector<int>> adj(numTasks); // Adjacency list: adj[i] is a list of tasks that depend on task i.
    vector<int> inDegree(numTasks, 0); // inDegree[i] is the number of prerequisites for task i.
    vector<int> result;             // The topological order of tasks.

    // 1. Build the graph and calculate in-degrees.
    for (const auto& pair : prerequisites) {
        int task = pair.first;      // Task that depends on the other.
        int prerequisite = pair.second; // Task that must be done first.
        adj[prerequisite].push_back(task); // Add edge: prerequisite -> task
        inDegree[task]++;             // Increment in-degree of the dependent task.
    }

    // 2. Find starting tasks (nodes with in-degree 0).
    queue<int> q;
    for (int i = 0; i < numTasks; ++i) {
        if (inDegree[i] == 0) {
            q.push(i); // Start with tasks that have no dependencies.
        }
    }

    // 3. Process tasks in topological order.
    while (!q.empty()) {
        int task = q.front();
        q.pop();
        result.push_back(task); // Add the task to the result.

        // For each task that depends on this task:
        for (int dependentTask : adj[task]) {
            inDegree[dependentTask]--; // Reduce its in-degree.
            if (inDegree[dependentTask] == 0) {
                // If its in-degree becomes 0, it can be done next.
                q.push(dependentTask);
            }
        }
    }

    // 4. Check for cycles.  If not all tasks were processed, there's a cycle.
    if (result.size() != numTasks) {
        return {}; // Return an empty vector to indicate a cycle.
    }

    return result;
}

// 2. Kahn's Algorithm with Custom Data (Course Scheduling with Credits)
//
//   -   Problem:  Similar to task scheduling, but each task (course) has a certain number of credits.  Find a valid schedule and calculate the total credits earned.
//   -   Real-world:  University course scheduling, where courses have prerequisites and credit hours.
//   -   Code:
struct Course {
    int id;
    int credits;
};

vector<int> topologicalSort_Kahn_Credits(int numCourses, const vector<pair<int, int>>& prerequisites, const vector<Course>& courses) {
    vector<vector<int>> adj(numCourses);
    vector<int> inDegree(numCourses, 0);
    vector<int> result;
    unordered_map<int, int> courseCredits; // Map from course ID to credits.

    for (const auto& course : courses) {
        courseCredits[course.id] = course.credits;
    }

    for (const auto& pair : prerequisites) {
        int course = pair.first;
        int prerequisite = pair.second;
        adj[prerequisite].push_back(course);
        inDegree[course]++;
    }

    queue<int> q;
    for (int i = 0; i < numCourses; ++i) {
        if (inDegree[i] == 0) {
            q.push(i);
        }
    }

    int totalCredits = 0; // Keep track of total credits.
    while (!q.empty()) {
        int courseId = q.front();
        q.pop();
        result.push_back(courseId);
        totalCredits += courseCredits[courseId]; // Add credits for the completed course.

        for (int dependentCourse : adj[courseId]) {
            inDegree[dependentCourse]--;
            if (inDegree[dependentCourse] == 0) {
                q.push(dependentCourse);
            }
        }
    }

    if (result.size() != numCourses) {
        return {};
    }
    cout << "Total Credits: " << totalCredits << endl;
    return result;
}

// 3. Kahn's Algorithm with Cycle Detection (Dependency Resolution with Error Handling)
//   -Problem: In a complex system, you need to resolve dependencies between components.  If there's a circular dependency, the system will fail.  Detect this case.
//   -Real world: Package management (like npm, apt), where packages can depend on each other.  A circular dependency ("package A depends on B, B depends on A") is a fatal error.
//   -Code:

vector<int> topologicalSort_Kahn_CycleDetection(int numNodes, const vector<pair<int, int>>& dependencies, bool& hasCycle) {
    vector<vector<int>> adj(numNodes);
    vector<int> inDegree(numNodes, 0);
    vector<int> result;
    hasCycle = false; // Initialize cycle detection flag.

    for (const auto& dependency : dependencies) {
        int node = dependency.first;
        int dependsOn = dependency.second;
        adj[dependsOn].push_back(node);
        inDegree[node]++;
    }

    queue<int> q;
    for (int i = 0; i < numNodes; ++i) {
        if (inDegree[i] == 0) {
            q.push(i);
        }
    }

    int visitedCount = 0; // Count of visited nodes.
    while (!q.empty()) {
        int node = q.front();
        q.pop();
        result.push_back(node);
        visitedCount++;

        for (int dependentNode : adj[node]) {
            inDegree[dependentNode]--;
            if (inDegree[dependentNode] == 0) {
                q.push(dependentNode);
            }
        }
    }

    if (visitedCount != numNodes) {
        hasCycle = true; // A cycle exists if not all nodes were visited.
        return {};      // Return empty vector to signal a cycle.
    }

    return result;
}

// 4. Kahn's Algorithm with Prioritization (Job Scheduling with Priorities)
//   -Problem:  You have jobs with dependencies, but also priorities.  Schedule the high-priority jobs first, while still respecting dependencies.
//   -Real-world:  Real-time operating systems (RTOS), where tasks have priorities.  The scheduler needs to run high-priority tasks as soon as their dependencies are met.
//   -Code:
struct Job {
    int id;
    int priority;
};

vector<int> topologicalSort_Kahn_Prioritized(int numJobs, const vector<pair<int, int>>& dependencies, const vector<Job>& jobs) {
    vector<vector<int>> adj(numJobs);
    vector<int> inDegree(numJobs, 0);
    vector<int> result;
    unordered_map<int, int> jobPriorities;

     for (const auto& job : jobs) {
        jobPriorities[job.id] = job.priority;
    }

    for (const auto& dependency : dependencies) {
        int job = dependency.first;
        int dependsOn = dependency.second;
        adj[dependsOn].push_back(job);
        inDegree[job]++;
    }

    // Use a priority queue (max-heap) to get highest priority job first.
    priority_queue<pair<int, int>> pq; // {priority, job_id}
    for (int i = 0; i < numJobs; ++i) {
        if (inDegree[i] == 0) {
            pq.push({jobPriorities[i], i}); // Push with priority.
        }
    }

    while (!pq.empty()) {
        int job = pq.top().second;
        pq.pop();
        result.push_back(job);

        for (int dependentJob : adj[job]) {
            inDegree[dependentJob]--;
            if (inDegree[dependentJob] == 0) {
                pq.push({jobPriorities[dependentJob], dependentJob});
            }
        }
    }

    if (result.size() != numJobs) {
        return {};
    }

    return result;
}

// 5. Kahn's Algorithm with Feedback (Interactive Task Management)
//   -Problem:  A user is performing tasks with dependencies.  Provide feedback:
//     -  Show the order in which tasks *can* be done.
//     -  If a task is blocked (due to unmet dependencies), tell the user *why*.
//   -Real-world:  Interactive task management tools, installation wizards, or guided workflows.
//   -Code:

vector<int> topologicalSort_Kahn_Feedback(int numTasks, const vector<pair<int, int>>& dependencies, vector<string>& feedbackMessages) {
    vector<vector<int>> adj(numTasks);
    vector<int> inDegree(numTasks, 0);
    vector<int> result;
    feedbackMessages.resize(numTasks); // Ensure enough space for feedback.

    for (const auto& dependency : dependencies) {
        int task = dependency.first;
        int dependsOn = dependency.second;
        adj[dependsOn].push_back(task);
        inDegree[task]++;
    }

    queue<int> q;
    for (int i = 0; i < numTasks; ++i) {
        if (inDegree[i] == 0) {
            q.push(i);
            feedbackMessages[i] = "Ready to start."; // Initial feedback.
        }
        else
        {
             feedbackMessages[i] = "Waiting for dependencies.";
        }
    }

    while (!q.empty()) {
        int task = q.front();
        q.pop();
        result.push_back(task);
        feedbackMessages[task] = "Completed.";

        for (int dependentTask : adj[task]) {
            inDegree[dependentTask]--;
            if (inDegree[dependentTask] == 0) {
                q.push(dependentTask);
                feedbackMessages[dependentTask] = "Ready to start."; // Dependency met.
            }
            else
            {
                 feedbackMessages[dependentTask] = "Waiting for dependencies.";
            }
        }
    }

    if (result.size() != numTasks) {
        return {};
    }
    return result;
}

int main() {
    // Example usage for each approach:
    cout << "1. Basic Kahn's Algorithm (Task Scheduling):" << endl;
    int numTasks1 = 6;
    vector<pair<int, int>> prerequisites1 = {{0, 1}, {0, 2}, {1, 3}, {2, 3}, {3, 4}, {3, 5}};
    vector<int> result1 = topologicalSort_Kahn_Basic(numTasks1, prerequisites1);
    if (!result1.empty()) {
        for (int task : result1) {
            cout << task << " ";
        }
        cout << endl;
    } else {
        cout << "Cycle detected!" << endl;
    }
    cout << endl;

    cout << "2. Kahn's Algorithm with Custom Data (Course Scheduling with Credits):" << endl;
    int numCourses2 = 4;
    vector<pair<int, int>> prerequisites2 = {{0, 1}, {2, 1}, {3, 0}};
    vector<Course> courses2 = {{0, 3}, {1, 2}, {2, 3}, {3, 1}};
    vector<int> result2 = topologicalSort_Kahn_Credits(numCourses2, prerequisites2, courses2);
    if (!result2.empty()) {
        for (int course : result2) {
            cout << course << " ";
        }
        cout << endl;
    } else {
        cout << "Cycle detected!" << endl;
    }
    cout << endl;

    cout << "3. Kahn's Algorithm with Cycle Detection (Dependency Resolution with Error Handling):" << endl;
    int numNodes3 = 4;
    vector<pair<int, int>> dependencies3 = {{0, 1}, {1, 2}, {2, 3}, {3, 0}}; // Introduce a cycle.
    bool hasCycle3;
    vector<int> result3 = topologicalSort_Kahn_CycleDetection(numNodes3, dependencies3, hasCycle3);
    if (!result3.empty()) {
        for (int node : result3) {
            cout << node << " ";
        }
        cout << endl;
    } else {
        if(hasCycle3)
            cout << "Cycle detected!" << endl;
        else
            cout << "No Cycle detected!" << endl;
    }
    cout << endl;

    cout << "4. Kahn's Algorithm with Prioritization (Job Scheduling with Priorities):" << endl;
    int numJobs4 = 5;
    vector<pair<int, int>> dependencies4 = {{0, 1}, {0, 2}, {1, 3}, {2, 3}, {3, 4}};
    vector<Job> jobs4 = {{0, 2}, {1, 1}, {2, 3}, {3, 1}, {4, 2}};
    vector<int> result4 = topologicalSort_Kahn_Prioritized(numJobs4, dependencies4, jobs4);
    if (!result4.empty()) {
        for (int job : result4) {
            cout << job << " ";
        }
        cout << endl;
    } else {
        cout << "Cycle detected!" << endl;
    }
    cout << endl;

    cout << "5. Kahn's Algorithm with Feedback (Interactive Task Management):" << endl;
    int numTasks5 = 6;
    vector<pair<int, int>> dependencies5 = {{0, 1}, {0, 2}, {1, 3}, {2, 3}, {3, 4}, {3, 5}};
    vector<string> feedbackMessages5;
    vector<int> result5 = topologicalSort_Kahn_Feedback(numTasks5, dependencies5, feedbackMessages5);
    if (!result5.empty()) {
        for (int task : result5) {
            cout << "Task " << task << ": " << feedbackMessages5[task] << endl;
        }
        cout << endl;
    } else {
        cout << "Cycle detected!" << endl;
    }
    cout << endl;

    return 0;
}

